### PR TITLE
IPv6 support for BGP peers in VRP devices

### LINKED ISSUE #11211: Some BGP sessions on Huawei no longer visible after 95820bf
**Please do not post bugs or ask questions**

**For help and support you can use our [Discord server](https://t.libren.ms/discord) or our [community site](https://community.librenms.org).**

Please confirm each of the sections below by putting an `x` in the box like `[x]`.

- [ ] Please provide **ALL** info asked for [here](http://docs.librenms.org/Support/FAQ/#faq20).
- [ ] Please provide as much detail as possible.
- [ ] Please do NOT post more than 10 lines of debug information here, use a pastebin service or GitHub Gists.

Please keep in mind that LibreNMS is a community project, it is run by people like you using their free time.
Which means it may take time before support is added to LibreNMS.

We encourage you to do it yourself and submit it in GitHub as a Pull request. It is not so difficult, most of
it can be done without writing a single line of code. You can find the documentation below :
👉  https://docs.librenms.org/Developing/os/Initial-Detection/

Love librenms? Please consider supporting our collective:
👉  https://t.libren.ms/donations

Hi folks,

Having an issue since 95820bf with IPv6 BGP session polling on Huawei VRP (hardware NE05E-SQ, software V300R003C10SPC500+V300R003SPH028 and V300R002C10SPC200). Discovery picks up v6 BGP sessions just fine, but polling never finds them.

```
$ ./validate.php
====================================
Component | Version
--------- | -------
LibreNMS  | 1.60-81-gcf023b936
DB Schema | 2020_02_10_223323_create_alert_location_map_table (159)
PHP       | 7.3.11-1+0~20191026.48+debian9~1.gbpf71ca0
MySQL     | 10.1.41-MariaDB-0+deb9u1
RRDTool   | 1.6.0
SNMP      | NET-SNMP 5.7.3
====================================

[OK]    Composer Version: 1.9.3
[OK]    Dependencies up-to-date.
[OK]    Database connection successful
[OK]    Database schema correct
[FAIL]  Discovery has not completed in the last 24 hours.
	[FIX]: 
	Check the cron job to make sure it is running and using discovery-wrapper.py
[WARN]  Your local git contains modified files, this could prevent automatic updates.
	[FIX]: 
	You can fix this with ./scripts/github-remove
	Modified Files:
	 rrd/.gitignore
$ 
```
(Discovery has definitely run in the previous 24 hours, not sure why that check fails).

From the poller debugger:

```
Checking BGP peer xxxx:xxxx::1 Fetching BGP4-MIB data... 
SNMP['/usr/bin/snmpget' '-v3' '-l' 'authPriv' '-n' "" '-a' 'SHA' '-A' 'PASSWORD' '-u' 'USER' '-x' 'AES' '-X' 'PASSWORD' '-OQUs' '-m' 'BGP4-MIB' '-M' '/opt/librenms/mibs:/opt/librenms/mibs/huawei' 'udp:HOSTNAME:161' 'bgpPeerState.xxxx:xxxx::1' 'bgpPeerAdminStatus.xxxx:xxxx::1' 'bgpPeerInUpdates.xxxx:xxxx::1' 'bgpPeerOutUpdates.xxxx:xxxx::1' 'bgpPeerInTotalMessages.xxxx:xxxx::1' 'bgpPeerOutTotalMessages.xxxx:xxxx::1' 'bgpPeerFsmEstablishedTime.xxxx:xxxx::1' 'bgpPeerInUpdateElapsedTime.xxxx:xxxx::1' 'bgpPeerLocalAddr.xxxx:xxxx::1']

Checking BGP peer x.x.x.1 Fetching BGP4-MIB data... 
SNMP['/usr/bin/snmpget' '-v3' '-l' 'authPriv' '-n' "" '-a' 'SHA' '-A' 'PASSWORD' '-u' 'USER' '-x' 'AES' '-X' 'PASSWORD' '-OQUs' '-m' 'BGP4-MIB' '-M' '/opt/librenms/mibs:/opt/librenms/mibs/huawei' 'udp:HOSTNAME:161' 'bgpPeerState.x.x.x.1' 'bgpPeerAdminStatus.x.x.x.1' 'bgpPeerInUpdates.x.x.x.1' 'bgpPeerOutUpdates.x.x.x.1' 'bgpPeerInTotalMessages.x.x.x.1' 'bgpPeerOutTotalMessages.x.x.x.1' 'bgpPeerFsmEstablishedTime.x.x.x.1' 'bgpPeerInUpdateElapsedTime.x.x.x.1' 'bgpPeerLocalAddr.x.x.x.1']
```
In an snmpwalk the v6 sessions don’t show up:

```
$ /usr/bin/snmpwalk -v3 -l authPriv -aSHA -A x -u x -xAES -X x -OQUs -mBGP4-MIB -M/opt/librenms/mibs:/opt/librenms/mibs/huawei router01 bgpPeerState
bgpPeerState.x.x.x.205 = established
bgpPeerState.x.x.x.1 = established
bgpPeerState.x.x.x.2 = established
bgpPeerState.x.x.x.3 = established
bgpPeerState.x.x.x.4 = established
bgpPeerState.x.x.x.6 = established
bgpPeerState.x.x.x.10 = established
bgpPeerState.x.x.x.100 = established
bgpPeerState.x.x.x.60 = established
bgpPeerState.x.x.x.68 = established
bgpPeerState.x.x.x.146 = established
$ 
```
I’d totally believe v6 sessions just not being exposed via SNMP with the new MIB. Is this a bug, or something I’m missing?

Thanks,

I.

### PR DESCRIPTION
**Pull request by PipoCanaja**
fixes #11211

 DO NOT DELETE THIS TEXT

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)
- [x] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

**Comment by PipoCanaja** [comment]
Code for BGP discovery is now extracted to its file, for readability. Other (TIMOS, etc) will be done later in a dedicated PR to avoid mixing stuff here.

File: LibreNMS/Util/IP.php [hunk 1]
```diff
@@ -73,6 +73,28 @@ public static function fromHexString($hex, $ignore_errors = false)
...
     }
 
+    /**
+     * Convert a decimal-string to an IP address. For example: "192 168 1 154" -> 192.168.1.254
+     * @param string $snmpOid
+     * @param bool $ignore_errors Do not throw exceptions, instead return null on error.
+     * @return IP|null
+     * @throws InvalidIpException
+     */
+    public static function fromSnmpString($snmpOid, $ignore_errors = false)
```
Comment (murrant) [inline LibreNMS/Util/IP.php]: Is this function applicable to more than one code location?  Does it work for both IPv4 and IPv6?

Comment (murrant) [reply LibreNMS/Util/IP.php]: https://github.com/librenms/ip-util

**Comment by murrant** [comment]
@PipoCanaja is there no snmp data that contains IPv6 peers?  Or another reason we don't see the test data change?

**Comment by PipoCanaja** [comment]
You are right, we did not have any ipv6 test-data up to now. Had to forge it based on the structure of a test file I received from community. Will see how it goes with it.

**Comment by murrant** [comment]
This pull request has been mentioned on **LibreNMS Community**. There might be relevant details there:

https://community.librenms.org/t/v1-63-release-changelog-april-2020/11828/1

**Pull request merged**